sdk/core/System.ClientModel/src/Internal/Logging/PipelineMessageLogger.cs

```diff
@@ -0,0 +1,258 @@
+                }
+                else
+                {
+                    ErrorResponseContent(_logger, requestId, content);
+                }
+            }
+        }
+        else
+        {
+            ClientModelEventSource.Log.ErrorResponseContent(requestId, content, textEncoding);
+        }
+    }
+
+    [LoggerMessage(LoggingEventIds.ErrorResponseContentEvent, LogLevel.Information, "Error response [{requestId}] content: {content}", SkipEnabledCheck = true, EventName = "ErrorResponseContent")]
+    private static partial void ErrorResponseContent(ILogger logger, string requestId, byte[] content);
+
+    [LoggerMessage(LoggingEventIds.ErrorResponseContentTextEvent, LogLevel.Information, "Error response [{requestId}] content: {content}", SkipEnabledCheck = true, EventName = "ErrorResponseContentText")]
+    private static partial void ErrorResponseContentText(ILogger logger, string requestId, string content);
+
+    public void LogErrorResponseContentBlock(string requestId, int blockNumber, byte[] content, Encoding? textEncoding)
+    {
+        if (_logger is not null)
+        {
+            if (_logger.IsEnabled(LogLevel.Information))
+            {
+                if (textEncoding != null)
+                {
+                    ErrorResponseContentTextBlock(_logger, requestId, blockNumber, textEncoding.GetString(content));
+                }
+                else
+                {
+                    ErrorResponseContentBlock(_logger, requestId, blockNumber, content);
+                }
+            }
+        }
+        else
+        {
+            ClientModelEventSource.Log.ErrorResponseContentBlock(requestId, blockNumber, content, textEncoding);
+        }
+    }
+
+    [LoggerMessage(LoggingEventIds.ErrorResponseContentBlockEvent, LogLevel.Information, "Error response [{requestId}] content block {blockNumber}: {content}", SkipEnabledCheck = true, EventName = "ErrorResponseContentBlock")]
+    private static partial void ErrorResponseContentBlock(ILogger logger, string requestId, int blockNumber, byte[] content);
+
+    [LoggerMessage(LoggingEventIds.ErrorResponseContentTextBlockEvent, LogLevel.Information, "Error response [{requestId}] content block {blockNumber}: {content}", SkipEnabledCheck = true, EventName = "ErrorResponseContentTextBlock")]
+    private static partial void ErrorResponseContentTextBlock(ILogger logger, string requestId, int blockNumber, string content);
+
+    #endregion
+
+    #region Helpers
+
+    private string FormatHeaders(IEnumerable<KeyValuePair<string, string>> headers)
+    {
+        var stringBuilder = new StringBuilder();
+        foreach (var header in headers)
+        {
+            stringBuilder.Append(header.Key);
+            stringBuilder.Append(':');
+            stringBuilder.Append(_sanitizer.SanitizeHeader(header.Key, header.Value));
+            stringBuilder.Append(Environment.NewLine);
```

Comment:
multi-line logs are very problematic - it's impossible to parse them or restore the full log record. 
We should not use them in anything new we create. Ideally we should pass a complex object to logger and let underlying formatter/provider decide how/if to serialize the array/map.

Reply:
Thank you for calling this out! Will update

Reply:
@lmolkova Do you know what the standard pattern would be for sanitizing the headers if I pass the headers directly? Passing something like this to use when formatting? https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.http.httpclientfactoryoptions.shouldredactheadervalue?view=net-9.0-pp#microsoft-extensions-http-httpclientfactoryoptions-shouldredactheadervalue

Reply:
I'm thinking like this maybe - https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingHttpMessageHandler.cs#L60